aten/src/ATen/native/ReduceOpsUtils.h

```diff
@@ -301,5 +301,72 @@ static void zero_numel_tensor_resize(Tensor& result, Tensor& result_indices,
   at::native::resize_output(result, sizes);
   at::native::resize_output(result_indices, sizes);
 }
+} // native
 
-}}  // at::native
+namespace meta {
+
+static void check_reduction_shape(
+    impl::MetaBase& meta,
```

Comment:
nit: I think it's cleaner to leave the meta object out of the helper. So the helper can just make various assertions (and return data if it needs to), and you can call `set_output()` or `build_reduce_op` directly inside of `TORCH_META_FUNC`